File: sdk/python/agentfield/harness/_result.py
from __future__ import annotations

from dataclasses import dataclass, field
from enum import Enum
from typing import Any, Dict, List, Optional


class FailureType(str, Enum):
    """Classifies how a harness invocation failed.

    Providers set this on RawResult so the runner can decide retry strategy:
    - ``none``: No failure.
    - ``crash``: Process killed by signal or non-zero exit with no output.
    - ``timeout``: Execution exceeded the time limit.
    - ``api_error``: Transient API-level error (rate limit, 5xx, etc.).
    - ``no_output``: Process exited OK but produced no output file.
    - ``schema``: Output file exists but fails schema validation.
    """

    NONE = "none"
    CRASH = "crash"
    TIMEOUT = "timeout"
    API_ERROR = "api_error"
    NO_OUTPUT = "no_output"
    SCHEMA = "schema"


@dataclass
class Metrics:
    duration_ms: int = 0
    duration_api_ms: int = 0
    num_turns: int = 0
    total_cost_usd: Optional[float] = None
    usage: Optional[Dict[str, Any]] = None
    session_id: str = ""
    # Token accounting parsed from the provider's result message. Left at 0 when
    # the provider's output does not expose token counts.
    input_tokens: int = 0
    output_tokens: int = 0
    cache_read_tokens: int = 0
    cache_creation_tokens: int = 0
    model: Optional[str] = None


@dataclass
class RawResult:
    result: Optional[str] = None
    messages: List[Dict[str, Any]] = field(default_factory=list)
    metrics: Metrics = field(default_factory=Metrics)
    is_error: bool = False
    error_message: Optional[str] = None
    failure_type: FailureType = FailureType.NONE
    returncode: Optional[int] = None


@dataclass
class HarnessResult:
    result: Optional[str] = None
    parsed: Any = None
    is_error: bool = False
    error_message: Optional[str] = None
    failure_type: FailureType = FailureType.NONE
    cost_usd: Optional[float] = None
    num_turns: int = 0
    duration_ms: int = 0
    session_id: str = ""
    messages: List[Dict[str, Any]] = field(default_factory=list)
    # Token accounting aggregated across all provider attempts. Zero when the
    # provider does not report token counts.
    input_tokens: int = 0
    output_tokens: int = 0
    cache_read_tokens: int = 0
    cache_creation_tokens: int = 0
    total_tokens: int = 0
    # Model slug reported by the provider (if any).
    model: Optional[str] = None

    @property
    def text(self) -> str:
        if self.result:
            return self.result
        return ""

File: sdk/go/harness/result.go
// Package harness dispatches tasks to external coding agents (CLI subprocesses)
// and extracts structured results with schema validation and retry logic.
package harness

// FailureType classifies how a harness invocation failed so the runner
// can decide on a retry strategy.
type FailureType string

const (
	FailureNone     FailureType = "none"
	FailureCrash    FailureType = "crash"
	FailureTimeout  FailureType = "timeout"
	FailureAPIError FailureType = "api_error"
	FailureNoOutput FailureType = "no_output"
	FailureSchema   FailureType = "schema"
)

// Metrics captured from a single provider execution.
type Metrics struct {
	DurationMS    int
	DurationAPIMS int
	NumTurns      int
	SessionID     string

	// CostUSD is the cost in USD reported by the provider for this single
	// execution, or nil when the provider does not report a cost. Mirrors the
	// Python SDK's Metrics.total_cost_usd (Optional[float]): nil means
	// "unknown", not "free". Only providers that surface a native cost (e.g.
	// Claude Code's result JSON) populate this; providers whose Python
	// counterparts derive cost via litellm token estimation leave it nil in Go
	// because there is no Go equivalent of litellm's pricing database.
	CostUSD *float64

	// Token accounting parsed from the provider's output. Left at zero when
	// the provider does not expose token counts. Mirrors the Python SDK's
	// Metrics token fields.
	InputTokens         int
	OutputTokens        int
	CacheReadTokens     int
	CacheCreationTokens int
}

// RawResult is the output from a single provider execution before schema
// parsing.
type RawResult struct {
	Result       string
	Messages     []map[string]any
	Metrics      Metrics
	IsError      bool
	ErrorMessage string
	FailureType  FailureType
	ReturnCode   int
}

// Result is the final harness output after schema validation, retries, and
// metrics accumulation.
type Result struct {
	// Result is the raw text output from the provider.
	Result string

	// Parsed holds the validated and deserialized structured output.
	// The caller passes a pointer to a struct; on success it is populated.
	Parsed any

	IsError      bool
	ErrorMessage string
	FailureType  FailureType

	// CostUSD is the total cost in USD accumulated across every provider
	// execution that contributed to this result (including failed retry
	// attempts), or nil when no execution reported a cost. Mirrors the Python
	// SDK's HarnessResult.cost_usd (Optional[float]).
	CostUSD *float64

	NumTurns   int
	DurationMS int
	SessionID  string
	Messages   []map[string]any

	// Token accounting aggregated across every provider execution that
	// contributed to this result (including failed retry attempts). Zero when
	// the provider does not report token counts. TotalTokens is
	// input+output — cache tokens are accounted separately. Mirrors the
	// Python SDK's HarnessResult token fields.
	InputTokens         int
	OutputTokens        int
	CacheReadTokens     int
	CacheCreationTokens int
	TotalTokens         int
}

// Text returns the result text, or empty string if nil.
func (r *Result) Text() string {
	return r.Result
}
